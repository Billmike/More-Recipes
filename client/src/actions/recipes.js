import axios from 'axios';
import {
  ADD_RECIPE,
  GET_RECIPES,
  GET_ONE_RECIPE,
  GET_USER_RECIPES,
  ADD_FAVORITE_RECIPE,
  FETCH_FAVORITE_RECIPES,
  EDIT_RECIPE,
  REMOVE_RECIPE,
  ADD_REVIEW,
  UPVOTE_RECIPE,
  DOWNVOTE_RECIPE
} from './types';
import './toastrConfig';

export const addRecipe = recipe => ({
  type: ADD_RECIPE,
  recipe
});

export const getAllRecipes = (recipes) => {
  return {
    type: GET_RECIPES,
    recipes,
  };
};

export const getOneRecipe = (recipe) => {
  return {
    type: GET_ONE_RECIPE,
    id: recipe.id,
    recipe
  };
};

export const getUserRecipe = (userRecipe) => {
  return {
    type: GET_USER_RECIPES,
    userRecipe,
  };
};

export const addFavorites = (favoriteRecipes) => {
  return {
    type: ADD_FAVORITE_RECIPE,
    favoriteRecipes
  };
};

export const fetchFavorites = (favoriteRecipes) => {
  return {
    type: FETCH_FAVORITE_RECIPES,
    favoriteRecipes,
  };
};

export const editRecipe = (id, updates) => ({
  type: EDIT_RECIPE,
  id,
  updates,
});

export const removeRecipe = ({ id } = {}) => ({
  type: REMOVE_RECIPE,
  id
});

export const addReview = (review) => {
  return {
    type: ADD_REVIEW,
    review,
  };
};

export const upVoteRecipe = (id) => {
  return {
    type: UPVOTE_RECIPE,
    id,
  };
};

export const downVoteRecipe = (id) => {
  return {
    type: DOWNVOTE_RECIPE,
    id,
  };
};

export const startGetAllRecipes = () => {
  return (dispatch) => {
    axios.get('http://localhost:8000/api/v1/recipes')
      .then((res) => {
        dispatch(getAllRecipes(res.data.recipeData));
      })
      .catch((error) => {
        console.log(error);
      });
  };
}

export const startAddRecipe = (recipeData = {}) => {
  return (dispatch) => {
    const {
      name = '',
      description = '',
      img_url = 'no-img-here',
      category = '',
      ingredients = '',
      instructions = ''
    } = recipeData;

    const recipe = {
      name,
      description,
      img_url,
      category,
      ingredients,
      instructions,
    };
    axios.post('http://localhost:8000/api/v1/recipes', recipe)
      .then(() => {
        toastr.success('Recipe added successfully.');
        dispatch(addRecipe({
          id: recipe.id,
          ...recipe
        }));
      });
  };
};

export const startGetUserRecipes = () => {
  return (dispatch) => {
    axios.get('http://localhost:8000/api/v1/users/recipes')
      .then((res) => {
        dispatch(getUserRecipe(res.data.userRecipe));
      })
      .catch((error) => {
        console.log(error);
      });
  };
};

export const startEditRecipe = (id, updates) => {
  return (dispatch) => {
    axios.put(`http://localhost:8000/api/v1/recipes/${id}/modify`, updates)
      .then(() => {
        toastr.success('Recipe edited successfully.');
        dispatch(editRecipe(id, updates));
      })
      .catch((error) => {
        console.log('_<_<_<_<_<__<_', error);
      });
  };
};

export const startRemoveRecipe = (id) => {
  return (dispatch) => {
    axios.delete(`http://localhost:8000/api/v1/recipes/${id}`)
      .then(() => {
        toastr.success('Recipe deleted successfully.');
        dispatch(removeRecipe(id));
      })
      .catch((error) => {
        console.log(error);
      });
  };
};

export const startGetOneRecipe = (id) => {
  return (dispatch) => {
    axios.get(`http://localhost:8000/api/v1/recipes/${id}`)
      .then((res) => {
        dispatch(getOneRecipe(res.data.recipeData));
      })
      .catch((error) => {
        console.log(error);
      });
  };
};

export const startAddFavoriteRecipes = (id) => {
  return (dispatch) => {
    axios.post(`http://localhost:8000/api/v1/recipes/${id}/favorites`)
      .then((res) => {
        toastr.success(res.data.message);
        return dispatch(addFavorites(res.data));
      })
      .catch((err) => {
        if (err.response.data
          .message === 'You need to be logged in to perform this action.') {
          toastr.error(err.response.data.message);
        } else if (err.response.data
          .message === 'Sorry, you cannot perform this action on your own recipe.') {
          toastr.error(err.response.data.message);
        }
      });
  };
};

export const startGetUserFavorites = (id) => {
  return (dispatch) => {
    axios.get(`http://localhost:8000/api/v1/users/${id}/favorites`)
      .then((res) => {
        dispatch(fetchFavorites(res.data));
      })
      .catch((err) => {
        console.log(err);
      });
  };
};

export const startUpvoteRecipe = (id) => {
  return (dispatch) => {
    axios.post(`http://localhost:8000/api/v1/recipes/${id}/votes/upvote`)
      .then((res) => {
        toastr.success(res.data.message);
      })
      .catch((err) => {
        if (err.response.data
          .message === 'You need to be logged in to perform this action.') {
          toastr.error(err.response.data.message);
        } else if (err.response.data
          .message === 'You already upvoted on this recipe') {
          toastr.warning(err.response.data.message);
        }
      });
  };
};

export const startDownVoteRecipe = (id) => {
  return (dispatch) => {
    axios.post(`http://localhost:8000/api/v1/recipes/${id}/votes/downvote`)
      .then((res) => {
        toastr.success(res.data.message);
      })
      .catch((err) => {
        if (err.response.data
          .message === 'You need to be logged in to perform this action.') {
          toastr.error(err.response.data.message);
        } else if (err.response.data
          .message === 'You already downvoted on this recipe') {
          toastr.warning(err.response.data.message);
        }
      });
  };
};

export const startAddReview = (id, data) => {
  return (dispatch) => {
    axios.post(`http://localhost:8000/api/v1/recipes/${id}/reviews`, data)
      .then((res) => {
        dispatch(addReview(res.data));
      })
      .catch((err) => {
        if (err.response.data
          .message === 'You need to be logged in to perform this action.') {
          toastr.error(err.response.data.message);
        } else if (err.response.data
          .message === 'You can\'t post an empty review. Please, enter a happy review for this recipe.') {
          toastr.warning(err.response.data.message);
        }
      });
  };
};
