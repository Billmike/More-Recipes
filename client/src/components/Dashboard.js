import React from 'react';
import { Link } from 'react-router-dom';
import pizza from '../assets/img/pizzza.jpg';
import '../assets/css/style.css';

const Dashboard = () => {
    return(
        <div className="container-fluid">
          <div className="row">
            <main className="container">
              <h1>My Recipes</h1>
        <div className="card-group">
          <div className="card">
            <a href="./recipe_detail.html"><img className="card-img-top" src={pizza} alt="Card image cap" /></a>
            <div className="card-body">
              <a href="./recipe_detail.html"><h4 className="card-title">Pizza Pepperoni with mushroom toppings.</h4></a>
              <p className="card-text">This is a wider card with supporting text below as a natural lead-in to additional content. This content is a little bit longer.</p>
              <a href="edit.html" className="btn btn-outline-success" role="button">Edit <i className="fa fa-pencil-square-o"></i></a> 
              <button type="button" className="btn btn-outline-danger" data-toggle="modal" data-target="#exampleModal">
                  Delete <i className="fa fa-trash-o" aria-hidden="true"></i>
                </button>
            </div>
            <div className="card-footer">
              <small className="text-muted">Last updated 3 mins ago</small>
            </div>
          </div>
          <div className="card">
            <a href="./recipe_detail.html"><img className="card-img-top" src={pizza} alt="Card image cap" /></a>
            <div className="card-body">
              <a href="./recipe_detail.html"><h4 className="card-title">Pizza Pepperoni with mushroom toppings.</h4></a>
              <p className="card-text">This is a wider card with supporting text below as a natural lead-in to additional content. This content is a little bit longer.</p>
              <a href="edit.html" className="btn btn-outline-success" role="button">Edit <i className="fa fa-pencil-square-o"></i></a> 
              <button type="button" className="btn btn-outline-danger" data-toggle="modal" data-target="#exampleModal">
                  Delete <i className="fa fa-trash-o" aria-hidden="true"></i>
                </button>
            </div>
            <div className="card-footer">
              <small className="text-muted">Last updated 3 mins ago</small>
            </div>
          </div>
          <div className="card">
            <a href="./recipe_detail.html"><img className="card-img-top" src={pizza} alt="Card image cap" /></a>
            <div className="card-body">
              <a href="./recipe_detail.html"><h4 className="card-title">Pizza Pepperoni with mushroom toppings.</h4></a>
              <p className="card-text">This is a wider card with supporting text below as a natural lead-in to additional content. This content is a little bit longer.</p>
              <a href="edit.html" className="btn btn-outline-success" role="button">Edit <i className="fa fa-pencil-square-o"></i></a> 
              <button type="button" className="btn btn-outline-danger" data-toggle="modal" data-target="#exampleModal">
                Delete <i className="fa fa-trash-o"></i>
              </button>
            </div>
            <div className="card-footer">
              <small className="text-muted">Last updated 3 mins ago</small>
            </div>
          </div>
        </div>
            </main>
          </div>
        </div>
    )
}

export default Dashboard;
