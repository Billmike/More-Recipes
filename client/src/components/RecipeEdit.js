import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import pizza from '../assets/img/pizzza.jpg';

const RecipeEdit = ({ id, name, description, category }) => {
	return (
		<div className="card-group">
        <div className="card index-card">
            <Link to={`/recipe/${id}`}><img className="card-img-top" src={ pizza } alt="Pizza" /></Link>
            <div className="card-body">
              <Link to={`/recipe/${id}`}><h4 className="card-title">{name}</h4></Link>
              <p className="card-text">{description}</p>
              <Link to={`/recipe/${id}`} className="btn border border-secondary rounded">45 <i className="fa fa-eye" aria-hidden="true" /></Link>
              <Link to={`/recipe/${id}`} className="btn border border-secondary rounded">  89 <i className="fa fa-heart" aria-hidden="true" /></Link>
              <Link to={`/recipe/${id}`} className="btn border border-secondary rounded"> 6 <i className="fa fa-comments" aria-hidden="true" /></Link>
            </div>
            <div className="container" style={{ textAlign : 'justify' }}>
            <Link to={`/edit/${id}`} className="btn btn-outline-success" role="button">Edit <i className="fa fa-pencil-square-o"></i></Link>
            <Link to={`/edit/${id}`} class="btn btn-danger" role="button">Delete <i class="fa fa-pencil-square-o"></i></Link>
            </div>
            <div className="card-footer">
              <small className="text-muted">Recipe Category - {category}</small>
            </div>
          </div>
          <div className="card index-card">
            <Link to={`/recipe/${id}`}><img className="card-img-top" src={ pizza } alt="Pizza" /></Link>
            <div className="card-body">
              <Link to={`/recipe/${id}`}><h4 className="card-title">{name}</h4></Link>
              <p className="card-text">{description}</p>
              <Link to={`/recipe/${id}`} className="btn border border-secondary rounded">45 <i className="fa fa-eye" aria-hidden="true" /></Link>
              <Link to={`/recipe/${id}`} className="btn border border-secondary rounded">  89 <i className="fa fa-heart" aria-hidden="true" /></Link>
              <Link to={`/recipe/${id}`} className="btn border border-secondary rounded"> 6 <i className="fa fa-comments" aria-hidden="true" /></Link>
            </div>
            <div className="container" style={{ textAlign : 'justify' }}>
            <Link to={`/edit/${id}`} class="btn btn-outline-success" role="button">Edit <i class="fa fa-pencil-square-o"></i></Link>
            <Link to={`/edit/${id}`} class="btn btn-danger" role="button">Delete <i class="fa fa-pencil-square-o"></i></Link>
            </div>
            <div className="card-footer">
              <small className="text-muted">Recipe Category - {category}</small>
            </div>
          </div>
          <div className="card index-card">
            <Link to={`/recipe/${id}`}><img className="card-img-top" src={ pizza } alt="Pizza" /></Link>
            <div className="card-body">
              <Link to={`/recipe/${id}`}><h4 className="card-title">{name}</h4></Link>
              <p className="card-text">{description}</p>
              <Link to={`/recipe/${id}`} className="btn border border-secondary rounded">45 <i className="fa fa-eye" aria-hidden="true" /></Link>
              <Link to={`/recipe/${id}`} className="btn border border-secondary rounded">  89 <i className="fa fa-heart" aria-hidden="true" /></Link>
              <Link to={`/recipe/${id}`} className="btn border border-secondary rounded"> 6 <i className="fa fa-comments" aria-hidden="true" /></Link>
            </div>
            <div className="container" style={{ textAlign : 'justify' }}>
            <Link to={`/edit/${id}`} class="btn btn-outline-success" role="button">Edit <i class="fa fa-pencil-square-o"></i></Link>
            <Link to={`/edit/${id}`} class="btn btn-danger" role="button">Delete <i class="fa fa-pencil-square-o"></i></Link>
            </div>
            <div className="card-footer">
              <small className="text-muted">Recipe Category - {category}</small>
            </div>
          </div>
      </div>
		)
};

export default connect()(RecipeEdit);
