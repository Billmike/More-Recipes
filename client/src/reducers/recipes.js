const recipeDefaultState = [];

export default (state = recipeDefaultState, action) => {
  switch (action.type) {
    case 'ADD_RECIPE':
      return [
        ...state,
        action.recipe,
      ];
    default:
      return state;
  }
};
