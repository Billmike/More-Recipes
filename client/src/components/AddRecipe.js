import React from 'react';
import RecipesForm from './RecipesForm';
import { connect } from 'react-redux';
import { addRecipe } from '../actions/recipes';

const AddRecipe = (props) => {
    console.log(props)
	return (
	   <div>
       <h1 className="container">Add Recipe</h1>
        <RecipesForm onSubmit={(recipe) => {
            props.dispatch(addRecipe(recipe));
            props.history.push('/')
        }} />
       </div>	
	)
}

export default connect()(AddRecipe);
