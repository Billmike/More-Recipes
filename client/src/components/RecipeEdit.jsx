import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import Modal from './Modal';
import { RemoveRecipeAction, GetUserRecipesAction } from '../actions/recipes';
import pizza from '../assets/img/pizzza.jpg';

class RecipeEdit extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selectedRecipe: undefined,
    };
    this.selectRecipe = this.selectRecipe.bind(this);
    this.handleClearSelectedRecipe = this.handleClearSelectedRecipe.bind(this);
    this.onRemoveRecipe = this.onRemoveRecipe.bind(this);
  }

  handleClearSelectedRecipe() {
    this.setState(() => ({
      selectedRecipe: undefined
    }));
  }

  selectRecipe() {
    this.setState(() => ({
      selectedRecipe: this.props.recipe.id
    }));
  }

  onRemoveRecipe() {
    this.props.RemoveRecipeAction(this.props.recipe.id).then(() => {
      this.props.GetUserRecipesAction();
    });
  }

  render() {
    return (
      <div className="col-md-4">
        <div className="card">
          <Link to={`/recipe/${this.props.recipe.id}`}>
            <img className="card-img-top" alt="Pizza" src={this.props.recipe.imageUrl} />
          </Link>
          <div className="card-body">
            <Link to={`/recipe/${this.props.recipe.id}`}>
              <h4 className="card-title">{this.props.recipe.name}</h4>
            </Link>
            <p className="card-text">{this.props.recipe.description}</p>
            <div className="group-btn">
              <div
                className="container action-btn"
                style={{ textAlign: 'justify' }}
              >
                <Link
                  to={`/edit/${this.props.recipe.id}`}
                  className="btn user-btn"
                  role="button"
                >
                  {' '}
                  <i className="far fa-edit brown" />{' '}
                </Link>
                <button
                  onClick={this.selectRecipe}
                  className="btn user-btn"
                  role="button"
                >
                  {' '}
                  <i className="fas fa-trash-alt crimson" />{' '}
                </button>
                <button
                  disabled={true}
                  className="btn user-btn user-btn-special"
                  role="button"
                >
                  {this.props.recipe.upVote}{' '}
                  <i className="fa fa-thumbs-up fa brown" />{' '}
                </button>
                <button
                  disabled={true}
                  className="btn user-btn user-btn-special"
                  role="button"
                >
                  {this.props.recipe.downVote}{' '}
                  <i className="fa fa-thumbs-down fa brown" />
                </button>
                <button
                  disabled={true}
                  className="btn user-btn user-btn-special"
                  role="button"
                >
                  {this.props.recipe.favorites}{' '}
                  <i className="fa fa-heart crimson" />
                </button>
              </div>
            </div>
          </div>
          <div className="card-footer pad-footer">
            <small className="text-muted small-text">
              Recipe Category - {this.props.recipe.category}
            </small>
          </div>
        </div>
        <Modal
          selectedRecipe={this.state.selectedRecipe}
          handleClearSelectedRecipe={this.handleClearSelectedRecipe}
          onRemoveRecipe={this.onRemoveRecipe}
        />
      </div>
    );
  }
}


const mapDispatchToProps = (dispatch) => {
  return {
    RemoveRecipeAction: recipeData => dispatch(RemoveRecipeAction(recipeData)),
    GetUserRecipesAction: () => dispatch(GetUserRecipesAction())
  };
};

export default connect(null, mapDispatchToProps)(RecipeEdit);
