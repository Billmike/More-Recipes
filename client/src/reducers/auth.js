import { isEmpty } from 'lodash';
import { SET_CURRENT_USER, GET_USER_INFORMATION } from '../actions/types';

export const initialState = {
  isAuthenticated: !!localStorage.getItem('authToken'),
  userDetails: {},
};

export default (state = initialState, action = {}) => {
  switch (action.type) {
    case SET_CURRENT_USER:
      return {
        ...state,
        isAuthenticated: !isEmpty(action.user),
        userDetails: action.user,
        user: action.user
      };
    case GET_USER_INFORMATION:
      return {
        ...state,
        isAuthenticated: !isEmpty(action.user),
        userDetails: action.user
      };
    default:
      return state;
  }
};
