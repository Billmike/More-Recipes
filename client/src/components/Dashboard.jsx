import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import Emoji from 'react-emoji-render';
import Footer from './Footer';
import RecipeEdit from './RecipeEdit';
import { GetUserRecipesAction } from '../actions/recipes';
import { getUserinfo } from '../actions/signinRequest';

/**
 * Component that renders the Dashboard page
 *
 * @class Dashboard
 *
 * @extends Component
 */
export class Dashboard extends Component {
  componentDidMount() {
    this.props.GetUserRecipesAction();
    this.props.getUserinfo();
  }

  render() {
    let availableRecipes;
    if (this.props.recipes.length > 0) {
      availableRecipes =
        this.props.recipes.map((recipe) => {
          return <RecipeEdit key={recipe.id} recipe={recipe} />;
        });
    } else {
      availableRecipes = (
        <p className="no-recipes-p">
          {this.props.user.username}, you do not have any recipes yet.{' '}
          <Link to="/add">Create one now.</Link>
          <Emoji text=":)" />
        </p>
      );
    }
    return (
      <div id="dashbordBody">
        <div className="container">
          <h2 className="dashboard-h2">
            {' '}
            Welcome to your Dashboard, {this.props.user.username}!
          </h2>
          <h4 className="dashboard-h4"> My Recipes </h4>
          <div className="row">{availableRecipes}</div>
        </div>
        <Footer />
      </div>
    );
  }
}

export const mapStateToProps = (state) => {
  return {
    recipes: state.recipes.userRecipe,
    user: state.auth.userDetails
  };
};

export default connect(
  mapStateToProps,
  { getUserinfo, GetUserRecipesAction }
)(Dashboard);
