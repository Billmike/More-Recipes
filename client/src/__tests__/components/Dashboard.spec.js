import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';
import { Dashboard, mapStateToProps } from '../../components/Dashboard';
import { RecipeEdit } from '../../components/RecipeEdit';
import recipes from '../fixtures/recipes';
import auth from '../fixtures/authUser';
import state from '../fixtures/state';

describe('<Dashboard />', () => {
  const GetUserRecipesAction = jest.fn();
  const getUserinfo = jest.fn();
  it('Should render correctly', () => {
    const wrapper = shallow(<Dashboard
      getUserinfo={() => Promise.resolve()}
      GetUserRecipesAction={() => Promise.resolve()}
      recipes={recipes[0]}
      user={auth} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('div').length).toBe(3);
  });
  it('Should render the RecipeEdit component as a child element', () => {
    const wrapper = shallow(<Dashboard
      getUserinfo={() => Promise.resolve()}
      GetUserRecipesAction={() => Promise.resolve()}
      recipes={recipes[0]}
      user={auth}
    >
      <div>
        <div className="container">
          <h2 className="dashboard-h2">
            Welcome to your Dashboard
          </h2>
          <h4 className="dashboard-h4"> My Recipes</h4>
          <div className="row">
            <RecipeEdit />
          </div>
        </div>
      </div>
    </Dashboard>);
  });
  it('Should call mapStateToProps', (done) => {
    mapStateToProps(state);
    done();
  });
});
