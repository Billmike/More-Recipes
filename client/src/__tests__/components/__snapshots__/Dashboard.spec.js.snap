// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> Should render correctly 1`] = `
<div
  id="dashbordBody"
>
  <div
    className="container"
  >
    <h2
      className="dashboard-h2"
    >
       
      Welcome to your Dashboard, 
      annabelle
      !
    </h2>
    <h4
      className="dashboard-h4"
    >
       My Recipes 
    </h4>
    <div
      className="row"
    >
      <p
        className="no-recipes-p"
      >
        annabelle
        , you do not have any recipes yet.
         
        <Link
          replace={false}
          to="/add"
        >
          Create one now.
        </Link>
        <Emoji
          text=":)"
        />
      </p>
    </div>
  </div>
  <Footer />
</div>
`;
